src/controller/InteractionPeer.h

```diff
@@ -0,0 +1,151 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/CommandSender.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/CHIPDeviceCallbacksMgr.h>
+#include <app/util/basic-types.h>
+#include <controller/data_model/zap-generated/CHIPClientCallbacks.h>
+
+namespace chip {
+namespace Controller {
+
+// The interface for the underlying connection of Interaction Peer
```

Comment:
The connection to a peer is a session.  Why can't we just use `CASESession` and `PASESession` objects to provide this abstraction?  These already implement the `Messaging::ExchangeDelegate` interface.  Isn't that fundamentally what's needed?

Reply:
Using the CASESession isn't enough since the `CASESession` class is more about authenticating a device with its certification. After that we need to call `SecureSessionMgr::NewPairing` which in turn calls `CASESession::DeriveSecureSession`. When talking to the device afterwards, we are not using the `CASESession` object but creating a new `ExchangeContext` instead.

Reply:
It's not the fact these objects require additional lifecycle prodding that precludes their use in the way I describe.  It's the fact that the stack does not directly use CASESession and PASESession objects to send messages.

Instead, the secure session manager provides a SessionHandle that can be used to construct exchange contexts, which then allows a key for matching messages traveling back and forth between the exchange manager and session manager singletons.  It's not how I would do it, but I see now this is the code does it.

Why then is SessionHandle not more fundamental to the design?  Why is this only accessible through the `InteractionPeerDelegate`?  I would think we would be storing this on construction of the `InteractionPeer`, or even extending from it, since it's the fundamental identifier of the InteractionPeer.  Why do we need to register a delegate to hold state for us?  Can it ever change?  If it did, would that be correct behavior?

And if we don't have an underlying Device object, what do we use as a delegate then?

Reply:
I believe you are correct and the centralized object is the `ExchangeContext`. `SessionHandle` is just a tuple used to create it. The challenge lies in two points:

* Our stack doesn't use `ExchangeContext` as the only handle to send and receive messages. The generated cluster code uses `deviceID`. The interaction model is using `Sessionhandle`. Moreover, we have `SendCommand` and `SendCommands` in `data_model/gen` and to my surprise they leads to different ExchangeContexts. This leads to the extra functions in ` InteractionPeerDelegate` like `GetDeviceID`, `GetSessionHandle` etc.
* The current `chip::Controller::Device` provides more functionality than a `ExchangeContext`. It will implicitly deserialize the object, setup the `CASESession` and even reset the whole transport layer. This leads to the `PreparePeer` function which does these work.

The `InteractionPeerDelegate` is designed so to make minimal changes to the code while decoupling the cluster client. However I believe these problems must be solved in the long run. What's the desired strategy to resolve these issues? Do we prefer to fix them once and for all in this PR or we prefer to solve it gradually?

Reply:
> The InteractionPeerDelegate is designed so to make minimal changes to the code while decoupling the cluster client. 

This is justification enough to me.

> However I believe these problems must be solved in the long run. What's the desired strategy to resolve these issues? Do we prefer to fix them once and for all in this PR or we prefer to solve it gradually?

I believe that if we can come to a cleaner concept of an unencrypted session and drive toward some other session-related cleanup, we may have additional opportunities for simplification and cleanup.

For now, this works and is an improvement.  I presume that if no other changes land and we want to unblock non-controller-initiated communication with this new code, we will simply need to provide something on the accessory side that can implement `InteractionPeerDelegate`.

Thanks for the background.